File: examples/py/connext.py
"""Basic RTI Connext DDS pub/sub wired into an asyncio TaskGroup. Requires
rti.connext, installed separately (not an SDK runtime dependency).
"""

import asyncio
import math

import rti.asyncio  # noqa: F401 - installs Connext's asyncio integration
import rti.connextdds as dds
import rti.types as idl

from rti_demo_ui import DemoUiApp


@idl.struct
class VehicleState:
    vehicle_id: str = ""
    x: float = 0.0
    y: float = 0.0
    heading: float = 0.0


async def main() -> None:
    app = DemoUiApp(title="Connext Demo")
    card = app.add_card("Fleet Telemetry")
    scene = card.add_scene_2d(
        width=600, height=400, bounds=(-100.0, 100.0, -100.0, 100.0)
    )
    scene.add_entity("vehicle-1", x=0.0, y=0.0)

    participant = dds.DomainParticipant(domain_id=0)
    topic = dds.Topic(participant, "VehicleStateTopic", VehicleState)
    writer = dds.DataWriter(topic)
    reader = dds.DataReader(topic)

    async def process_samples() -> None:
        async for data in reader.take_data_async():
            scene.update_entity(
                data.vehicle_id, x=data.x, y=data.y, heading=data.heading
            )

    async def publish_samples() -> None:
        angle = 0.0
        while True:
            angle += 0.05
            sample = VehicleState(
                vehicle_id="vehicle-1",
                x=40.0 * math.cos(angle),
                y=40.0 * math.sin(angle),
                heading=math.degrees(angle) % 360,
            )
            writer.write(sample)
            await asyncio.sleep(0.1)

    try:
        async with asyncio.TaskGroup() as tasks:
            tasks.create_task(app.run())
            tasks.create_task(process_samples())
            tasks.create_task(publish_samples())
    except asyncio.CancelledError:
        pass
    finally:
        await app.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/cpp/connext.cpp
/*
 * (c) 2026 Copyright, Real-Time Innovations, Inc.  All rights reserved.
 *
 * RTI grants Licensee a license to use, modify, compile, and create derivative
 * works of the Software.  Licensee has the right to distribute object form
 * only for use with RTI products.  The Software is provided "as is", with no
 * warranty of any type, including any warranty for fitness for any purpose.
 * RTI is under no obligation to maintain or support the Software.  RTI shall
 * not be liable for any incidental or consequential damages arising out of the
 * use or inability to use the software.
 */

// Basic RTI Connext DDS pub/sub wired in through the application lifecycle
// described in docs/architecture.md. Built only when BUILD_CONNEXT_EXAMPLE=ON.
#include <atomic>
#include <chrono>
#include <cmath>
#include <dds/dds.hpp>
#include <rti/rti.hpp>
#include <rti_demo_ui/rti_demo_ui.hpp>
#include <thread>

#include "VehicleState.hpp"
#include "console_control.hpp"

int main() {
    using namespace rti::demo::ui;

    rti_demo_ui_examples::ConsoleControl control;
    DemoUiApp app("Connext Demo");
    control.start([&app]() { app.stop(); });
    Card* card = app.add_card("Fleet Telemetry");
    Scene2DViewport* scene =
        card->add_scene_2d(600, 400, {-100.0, 100.0, -100.0, 100.0});
    scene->add_entity("vehicle-1", 0.0, 0.0);

    // Declared in parent-to-child order so reverse destruction is safe.
    dds::domain::DomainParticipant participant(0);
    dds::topic::Topic<VehicleState> topic(participant, "VehicleStateTopic");
    dds::pub::DataWriter<VehicleState> writer(topic);
    dds::sub::DataReader<VehicleState> reader(topic);

    rti::sub::SampleProcessor processor;
    processor.attach_reader(
        reader, [&scene](const rti::sub::LoanedSample<VehicleState>& sample) {
            if (!sample.info().valid()) return;
            const auto& data = sample.data();
            scene->update_entity(data.vehicle_id, data.x, data.y, data.heading);
        });

    std::atomic<bool> stop_writer{false};
    std::thread writer_thread([&writer, &stop_writer]() {
        double angle = 0.0;
        while (!stop_writer.load()) {
            VehicleState sample;
            sample.vehicle_id = "vehicle-1";
            angle += 0.05;
            sample.x = 40.0 * std::cos(angle);
            sample.y = 40.0 * std::sin(angle);
            sample.heading = angle * 180.0 / M_PI;
            writer.write(sample);
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    });

    app.run();

    stop_writer.store(true);
    writer_thread.join();
    control.finish();
    return 0;
}
